types/artifact.go

```diff
@@ -82,24 +90,26 @@ type ArtifactInfo struct {
 // BlobInfo is stored in cache
 type BlobInfo struct {
 	SchemaVersion     int
-	Digest            string             `json:",omitempty"`
-	DiffID            string             `json:",omitempty"`
-	OS                *OS                `json:",omitempty"`
-	PackageInfos      []PackageInfo      `json:",omitempty"`
-	Applications      []Application      `json:",omitempty"`
-	Misconfigurations []Misconfiguration `json:",omitempty"`
-	OpaqueDirs        []string           `json:",omitempty"`
-	WhiteoutFiles     []string           `json:",omitempty"`
-	Size              int                `json:",omitempty"`
+	Digest            string                       `json:",omitempty"`
+	DiffID            string                       `json:",omitempty"`
+	OS                *OS                          `json:",omitempty"`
+	PackageInfos      []PackageInfo                `json:",omitempty"`
+	Applications      []Application                `json:",omitempty"`
+	Misconfigurations []Misconfiguration           `json:",omitempty"`
+	OpaqueDirs        []string                     `json:",omitempty"`
+	WhiteoutFiles     []string                     `json:",omitempty"`
+	Size              int                          `json:",omitempty"`
+	CustomResources   map[string]AnalyzerResources `json:",omitempty"`
```

Comment:
How about this?

```suggestion
	CustomResources   []CustomResource `json:",omitempty"`
```

We already have the `Type` field in CustomResource. So, we can marshal `Info` depending on the type. Then, we can merge these resources based on FilePath + Type.

```
		for _, customResource := range layer.CustomResources {	
			// e.g. key => /tmp/something/malicious/custom:malware
			key := fmt.Sprintf("%s/custom:%s", customResource.FilePath, customResource.Type)
			nestedMap.SetByString(key, sep, customResource)
		}
```

Also, we have to replace the keys of PackageInfos, Applications, and Configs.
https://github.com/aquasecurity/fanal/blob/048c06dd1202dabca54ac99207194d5732fe9bd6/applier/docker.go#L85-L97

Like
```
nestedMap.SetByString(fmt.Sprintf("%s/type:ospkg", pkgInfo.FilePath), sep, pkgInfo)
```

Then, you can write as below

```
	_ = nestedMap.Walk(func(keys []string, value interface{}) error {
		switch v := value.(type) {
		case types.PackageInfo:
			mergedLayer.Packages = append(mergedLayer.Packages, v.Packages...)
		case types.Application:
			mergedLayer.Applications = append(mergedLayer.Applications, v)
		case types.Config:
			mergedLayer.Configs = append(mergedLayer.Configs, v)
		case types.CustomResource:
			mergedLayer.CustomResources = append(mergedLayer.CustomResources, v)
		}
		return nil	})
```

Reply:
Yes, As the CustomResource schema is changed, Type can help us differentiate the custom resource. I have updated that follows this design.